### ISSUE TITLE
`icon` and `icon-end` attributes as shorthands for slotting an icon into `prefix` and `suffix` slots

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#234]
This is blocked on:
- #430 (otherwise how would people set these properties on the icons?)

----

Currently we use `prefix` and `suffix` **slots**. This is nice for making complex things possible (you can slot anything, not just icons!), but not for making simple things easy (the use case of slotting icons is *incredibly* common).

This is a lot to type just to get an icon:
```html
<wa-button>
  <wa-icon slot="prefix" name="gear" variant="solid"></wa-icon>
  Settings
</wa-button>
```

In the spirit of _Syntax should not require more than the minimum to specify author intent_, would it be possible to have `icon` and `icon-end` attributes that are sugar for this (and use the default resolution for `<wa-icon>`)? These would make the icon button use case a lot nicer (see [#164](https://github.com/shoelace-style/webawesome/issues/164#issuecomment-2486207179).

Compare the code above to:

```html
<wa-button icon="gear">Settings</wa-button>
```

There should also be a special value to **disable** the icon. That way, certain components could _default_ to having icons (callout comes to mind). The ideal syntax seems `icon="none"` or `icon="false"`, but can we guarantee that no icon library will have that? Perhaps it can be simply "any value that does not resolve to an icon disables the icon", then `none` and `false` will work for 99% of cases, and folks have options for that 1%.

This affects several components: 
- [ ] `<wa-button>`, 
- [ ] `<wa-select>`
- [ ] `<wa-input>`
- [ ] `<wa-option>`
- [ ] `<wa-menu-item>`
- [ ] `<wa-breadcrumb-item>`
- [ ] `<wa-radio-button>`
- [ ] `<wa-callout>`
- [ ] `<wa-badge>` (see #254)
- [ ] `<wa-tag>` (see #254)
- [ ] `<wa-rating>`
- [ ] `<wa-tree-item>`

**Comment by claviska** [comment]
The example code only accounts for icon name, not family or variant (new to WA, not in Shoelace).

Those _could_ be extra attributes, of course.

**Comment by LeaVerou** [comment]
> The example code only accounts for icon name, not family or variant (new to WA, not in Shoelace).
> 
> Those _could_ be extra attributes, of course.

Shortcuts are a way to make common things easy. It’s okay if a shortcut doesn't do everything, because you can always use the non-shortcut syntax.

That said, I wonder how frequently you want to set variant or family on a case by case basis rather than once for the whole page. Anecdotally, while there are one-off overrides (mainly between stroked and solid), there are usually very clear defaults in each project. But that's for`<wa-icon>`'s UX audit.

**Comment by claviska** [comment]
As a maintainer, it shows a willingness to use a pattern and the expectation is we’re open to expanding it. Unsolicited PRs will just come. It’s a slippery slope is all.

> That said, I wonder how frequently you want to set variant or family on a case by case basis rather than once for the whole page.

Family/variant is typically set once per page, except for one-offs for icons that don’t exist in the library. Or if you’re using solid and outlines icons on the same page, which is somewhat common.

**Comment by LeaVerou** [comment]
> As a maintainer, it shows a willingness to use a pattern and the expectation is we’re open to expanding it. Unsolicited PRs will just come. It’s a slippery slope is all.

We don't have to accept every PR if we don’t feel it’s an improvement on our API. 

> Family/variant is typically set once per page, except for one-offs for icons that don’t exist in the library. Or if you’re using solid and outlines icons on the same page, which is somewhat common.

Thanks, that’s what I suspected. I'll open a separate issue so we can discuss the ergonomics of `<wa-icon>`.

**Comment by claviska** [comment]
While I appreciate the proposed syntax being shorter, the more I think about this, the more it bothers me to pollute a dozen or more components with icon-specific attributes and logic. In my opinion, I don't think we should pursue this approach to icons.

**Comment by claviska** [comment]
Another thought came up — some components, like button, have multiple slots for icons. What does the syntax look for this?

```html
<wa-button>
  <wa-icon slot="prefix" name="link" variant="solid"></wa-icon>
  Settings
  <wa-icon slot="prefix" name="arrow-up-right-from-square" variant="solid"></wa-icon>
</wa-button>
```

And if the answer is "the shorthand works for many uses cases but not all" then I'd warn that this is exactly the type of thing that becomes very hard to [defend against](https://github.com/shoelace-style/webawesome/issues/234#issuecomment-2489870074), and the next thing we know every component has dozens of unrelated icon attributes polluting its APIs.

I think it's perfectly to ask users to slot in an icon where an icon is needed and I really don't like the idea of mixing this logic into so many components.

**Comment by claviska** [comment]
After discussing this in a call, it sounds like @LeaVerou is OK with scrapping this at least for now.

**Comment by LeaVerou** [comment]
> After discussing this in a call, it sounds like [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/175836?u=af5ecaec009f4cf22e5978db7ba974c19585d459&amp;v=4&amp;size=80">@LeaVerou](https://github.com/LeaVerou) is OK with scrapping this at least for now.

FWIW I still _strongly_ believe it's the right direction, but it's not on the critical path for anything we want to ship. It would make a lot of code much nicer, but so would other things.

> Another thought came up — some components, like button, have multiple slots for icons. What does the syntax look for this?
> 
> <wa-button>
>   <wa-icon slot="prefix" name="link" variant="solid"></wa-icon>
>   Settings
>   <wa-icon slot="prefix" name="arrow-up-right-from-square" variant="solid"></wa-icon>
> </wa-button>

I see this as no different than having e.g. a `hint` attribute and a `hint` slot. The slot makes complex things possible and the attribute makes simple things easy. **Together they give you the holy grail of API design.**

We didn't add additional `hint-*` attributes to customize the hint, we just direct people to use the slot. The hassle of writing out a whole `<wa-icon>` feels worth it when you're doing something out of the ordinary, but it feels like gruntwork when you have to do it for every icon you add (and adds a lot of noise to the markup).

Shortcuts are one of the most common ways to improve the ease-of-use to complexity curve of an API when the complex feature has a high concentration of a certain type of use case, and are very widely used in the web platform as well. This seems like the perfect candidate for a shortcut, since I'd estimate > 90% of prefix/suffix use cases are icons (possibly even > 99%).

That said, looking at your particular example, did you mean to have both icons slotted to `prefix`? That would look pretty weird, no?

<img width="192" alt="Image" src="https://github.com/user-attachments/assets/e4f3dca6-81c0-4dde-b1e6-3342b6e51a49" />

If the second icon was meant to be slotted to `suffix`, I don't see the issue? That’s what `icon-end` is for.

> And if the answer is "the shorthand works for many uses cases but not all" then I'd warn that this is exactly the type of thing that becomes very hard to [defend against](https://github.com/shoelace-style/webawesome/issues/234#issuecomment-2489870074), and the next thing we know every component has dozens of unrelated icon attributes polluting its APIs.
> 
> I think it's perfectly to ask users to slot in an icon where an icon is needed and I really don't like the idea of mixing this logic into so many components.

Regardless of this particular issue, we should be *very* conservative around accepting outside contributions that change the API shape. We put a lot of effort into designing our APIs, and should not lower our standards to be nice to contributors. We can always work with them to improve API shape if the feature is something we do want to do, but if not, we have every right not to accept the PR, and we should communicate this clearly in the contribution guide, prompting folks to check with us before implementing things that change the API shape. Feel free to direct people to me if you have trouble saying no, I promise I’ll be nice 😁 

Btw this is also why I think there may be value in publicly posting mini "specs" of features we have consensus to implement, but don't have time to prioritize. This way, those with the energy to help out know what API shape to follow. Or perhaps we could publish a list of these things, and prompt people to contact us for the API shape.

-----

> While I appreciate the proposed syntax being shorter, the more I think about this, the more it bothers me to pollute a dozen or more components with icon-specific attributes and logic. In my opinion, I don't think we should pursue this approach to icons.

Is this a comment about the API or about the codebase architecture? If the latter, would it help to start from components that already have a dependency on `<wa-icon>`? Or abstract the logic away into a single module that these components import?

**Comment by claviska** [comment]
> Btw this is also why I think there may be value in publicly posting mini "specs" of features we have consensus to implement, but don't have time to prioritize.

I tend to grasp new ideas better through seeing code so, instead of speccing things out with words, a handful of code examples demonstrating how it's intended to work will appeal more to me. Then, once we're all on the same page, we can shape it into words and make it a convention, e.g. in contributing.md or somewhere relevant.

**Comment by claviska** [comment]
> Is this a comment about the API or about the codebase architecture?

For lack of better words, it feels disgusting that so many components would have so many icon attributes. It's just dirty, IMO.

**Comment by LeaVerou** [comment]
> > Btw this is also why I think there may be value in publicly posting mini "specs" of features we have consensus to implement, but don't have time to prioritize.
> 
> I tend to grasp new ideas better through seeing code so, instead of speccing things out with words, a handful of code examples demonstrating how it's intended to work will appeal more to me. Then, once we're all on the same page, we can shape it into words and make it a convention, e.g. in contributing.md or somewhere relevant.

Good specs have examples too. 😊 

> > Is this a comment about the API or about the codebase architecture?
> 
> For lack of better words, it feels disgusting that so many components would have so many icon attributes. It's just dirty, IMO.

I still don’t get it — dirty in terms of codebase architecture or API?
Would it be more palatable to only do those that already depend on `<wa-icon>`, such as button or callout (at least at first)?

**Comment by claviska** [comment]
> I still don’t get it — dirty in terms of codebase architecture or API?

Both.

> Would it be more palatable to only do those that already depend on <wa-icon>, such as button or callout (at least at first)?

No, but that raises a good point. Any component that would support such attributes would need to make icon a dependency to guarantee its existence.
